### ISSUE TITLE
Seems to be vulnerable to the pickle security issue (remote code execution)

### ISSUE THREAD

**Issue by robvdl** [ericrasmussen/pyramid_redis_sessions#53]
I am using Pyramid and Beaker at the moment for sessions, but Beaker is currently also vulnerable to the pickle issue found in so many Python frameworks:

http://www.balda.ch/posts/2013/Jun/23/python-web-frameworks-pickle/

Django has fixed it now, but Beaker and Pyramid's built-in session haven't

I was hoping that pyramid_redis_sessions would not be vulnerable to this security vulverability, but a quick grep of the code shows, that it too is using pickle instead of json for serialisation.

If this is an issue, can we please get this fixed by using JSON for serialisation, not pickle

**Comment by joulez** [comment]
pickle via `PickleSerializer` is the default in pyramid as well. Because it's possible that third party apps/libs could possibly break changing the hardwired default (there is limits to what json can serialize). The better option would be to at least add a json serializer and perhaps issue deprecation warnings is relation to the default changing in future.

**Comment by mmerickel** [comment]
This comes up every once in a while and the answer is always the same. Yes, pickle is insecure. That's why we sign the content. If your secret is compromised then you're screwed, obviously. If your secret is not compromised then I have yet to see an attack against this technique. This is what Pyramid does by default, and beaker for years before it and I've never heard of an exploit.

**Comment by robvdl** [comment]
I agree, but I think it only becomes a problem when combined with another attack (such as a directory traversal bug in a different library), as a security review company have told me in the past.  It's highly unlikely I agree, but then it's not entirely impossible either they tell me.  I am not sure how convinced I am about this myself, but I see their point.

Personally I don't see much use in storing anything other than dicts, lists, int, bool, and string types in the session and not complex objects, so for most cases a JSON serializer is going to be just fine, but I understand that other people may already be storing objects in their session.

**Comment by redspider** [comment]
Not expressing an opinion on solutions, however it seemed important to note the following:

1. The pickling/unpickle of storage within redis itself is probably not an issue, the ability for an attacker to modify the blob inside redis is predicated on their ability to access redis as well as know the secret key. At this point I think it's reasonable to say the application is well and truly compromised, typically applications are not built to defend themselves from a compromise of their backend data storage.

2. However, the key for the particular redis blob that is to be decoded is stored in the public facing cookie. This is done using pickle as well, via pyramid.session/signed_*serialize which makes use of a sha1 hmac for signing.

3. There is no way for an application to modify the behaviour of signed_*serialize other than monkeypatching, nor is it possible to instruct redis_sessions to use a different serialization option, so typical app developers cannot mitigate this by switching serialization methods.

A proposal to replace the hardcoded serialization exists here: https://github.com/ericrasmussen/pyramid_redis_sessions/issues/38 - webob.SignedCookieProfile uses sha512 and json serialization by default.

**Comment by jgelens** [comment]
@ericrasmussen 

I guess using json will work fine.
I've tried:
redis.sessions.serialize = json.dumps
redis.sessions.deserialize = json.loads

But the code expects json to be imported. How should this be used?

**Comment by ericrasmussen** [comment]
@jgelens pyramid_redis_sessions uses pyramid.config's maybe_dotted function (http://docs.pylonsproject.org/docs/pyramid/en/latest/api/config.html) to resolve the import. I tested those settings as you have them and they worked fine. If they didn't work for you, can you open a separate issue with a reproducible example? Thanks!

**Comment by ericrasmussen** [comment]
FYI, I'm closing this issue for now because pickle support is required by the pyramid ISession interface (http://docs.pylonsproject.org/projects/pyramid/en/latest/api/interfaces.html#pyramid.interfaces.ISession), and because users can choose other serializers if needed.

Thanks everyone for the discussion and feedback -- I agree with the concerns/issues and I'm all for getting people to use a safer form of data serialization, but I also can't break pyramid support (the whole reason this library exists!) for the default.

**Comment by jvanasco** [comment]
This issue has been addressed in the 1.5.1 version of my fork, https://github.com/jvanasco/pyramid_session_redis

The problematic call to `pickle` has nothing to do with `ISession`.  Serializing the data with pickle is absolutely safe in this context of server-side sessions.  Changing the serializer to JSON as suggested above will not address this issue.

The potential vulnerability in this library is from signing/reading the cookie value via `pyramid.session`s deprecated `signed_serialize` and `signed_deserialize` functions. Those functions call `pickle.loads` on the user-submitted cookie content (if the hmac secret is broken). The value of this payload is just the backend session_id.

The potential fixes are:

* don't sign the cookie payload, and remove the calls to these functions
* rework the code to do what Pyramid does for signed cookies, and invoke webob's signing functions. this is what I did, with a small difference.  Pyramid uses JSON serialization on the cookie value, but since the session_id is an alphanumeric string, I just wrote a null-serializer that mimics the interface and passes the data through.








https://github.com/ericrasmussen/pyramid_redis_sessions/blob/6888d9d2ba530a4a73ee35de009b812f8f32d311/pyramid_redis_sessions/__init__.py#L284

https://github.com/ericrasmussen/pyramid_redis_sessions/blob/6888d9d2ba530a4a73ee35de009b812f8f32d311/pyramid_redis_sessions/__init__.py#L260
